omegaconf/grammar_parser.py

```diff
@@ -19,7 +19,8 @@
 # Build regex pattern to efficiently identify typical interpolations.
 # See test `test_match_simple_interpolation_pattern` for examples.
 _config_key = r"[$\w]+"  # foo, $0, $bar, $foo_$bar123$
-_node_path = f"(\\.)*({_config_key}(\\.{_config_key})*)?"  # foo, .foo.$bar
+_node_access = f"\\.{_config_key}|\\[{_config_key}\\]"  # .foo, [foo], [$bar]
+_node_path = f"(\\.)*({_config_key}({_node_access})*)?"  # foo, .foo.$bar, foo[bar]
```

Comment:
Can you add another long string to the benchmark that contains this style of key to see performance is good?
e.g:
```
("$a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1]}", ValueKind.INTERPOLATION),
```

Reply:
Done in c64d446 => looks good to me (first two rows):
```
ame (time in us)                                                                                                     Min                    Max                Mean              StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_get_value_kind[${a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1]}-ValueKind.INTERPOLATION-True]        6.0050 (1.87)        122.2720 (1.70)       6.7207 (1.87)       2.2021 (1.60)       6.5400 (1.87)      0.1320 (2.49)     467;5170      148.7930 (0.54)      52629           1
test_get_value_kind[${a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1].a[1]}-ValueKind.INTERPOLATION-False]       3.2710 (1.02)        111.6820 (1.55)       3.6911 (1.03)       1.5701 (1.14)       3.5760 (1.02)      0.0920 (1.74)    519;13440      270.9206 (0.98)     103897           1
test_get_value_kind[${a:b,c,d}-ValueKind.INTERPOLATION-False]                                                      3.2090 (1.0)         116.0530 (1.62)       3.6442 (1.01)       2.1572 (1.57)       3.5210 (1.00)      0.0960 (1.81)     609;5267      274.4071 (0.99)     104385           1
test_get_value_kind[${long_string1xxx}_${long_string2xxx:${key}}-ValueKind.INTERPOLATION-False]                    3.2320 (1.01)         71.8260 (1.0)        3.5997 (1.0)        1.4992 (1.09)       3.5240 (1.01)      0.1190 (2.25)     508;1239      277.7977 (1.0)      100868           1
test_get_value_kind[simple-ValueKind.VALUE-False]                                                                  3.2540 (1.01)         96.0400 (1.34)       3.6567 (1.02)       1.6797 (1.22)       3.5450 (1.01)      0.0620 (1.17)     540;6027      273.4733 (0.98)      97638           1
test_get_value_kind[simple-ValueKind.VALUE-True]                                                                   3.2690 (1.02)         88.2840 (1.23)       3.6320 (1.01)       1.7461 (1.27)       3.5060 (1.0)       0.1080 (2.04)      118;427      275.3303 (0.99)      18425           1
test_get_value_kind[${a:${b}}-ValueKind.INTERPOLATION-False]                                                       3.3010 (1.03)         87.7480 (1.22)       3.6397 (1.01)       1.4013 (1.02)       3.5710 (1.02)      0.0840 (1.58)     369;6208      274.7492 (0.99)      98252           1
test_get_value_kind[${${b}}-ValueKind.INTERPOLATION-False]                                                         3.3310 (1.04)        254.8840 (3.55)       3.6705 (1.02)       1.8336 (1.33)       3.5620 (1.02)      0.0630 (1.19)     472;6240      272.4389 (0.98)     100091           1
test_get_value_kind[${a}-ValueKind.INTERPOLATION-False]                                                            3.3660 (1.05)         74.4900 (1.04)       3.6511 (1.01)       1.3773 (1.0)        3.5700 (1.02)      0.0530 (1.0)      130;1378      273.8892 (0.99)      35797           1
test_get_value_kind[${a}-ValueKind.INTERPOLATION-True]                                                             4.3350 (1.35)         90.1880 (1.26)       4.7217 (1.31)       1.6762 (1.22)       4.5880 (1.31)      0.0860 (1.62)     345;1835      211.7900 (0.76)      47697           1
test_get_value_kind[${a:b,c,d}-ValueKind.INTERPOLATION-True]                                                       4.7510 (1.48)         99.3830 (1.38)       5.3331 (1.48)       1.9930 (1.45)       5.1620 (1.47)      0.1160 (2.19)     580;4874      187.5073 (0.67)      54958           1
test_get_value_kind[${${b}}-ValueKind.INTERPOLATION-True]                                                        249.8020 (77.84)       435.5180 (6.06)     260.7636 (72.44)     26.3446 (19.13)    252.9805 (72.16)     3.1620 (59.66)       14;17        3.8349 (0.01)        166           1
test_get_value_kind[${a:${b}}-ValueKind.INTERPOLATION-True]                                                      333.7430 (104.00)   10,315.0600 (143.61)   387.6692 (107.69)   546.8473 (397.06)   338.8680 (96.65)    35.7640 (674.79)       2;16        2.5795 (0.01)        334           1
test_get_value_kind[${long_string1xxx}_${long_string2xxx:${key}}-ValueKind.INTERPOLATION-True]                   530.4720 (165.31)    1,142.6960 (15.91)    583.2386 (162.02)    67.1833 (48.78)    562.0560 (160.31)   14.7300 (277.92)      23;77        1.7146 (0.01)        270           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```